--- Racer.gcl	2022-10-29 10:36:50.744012847 -0400
+++ Racerx.gcl	2022-10-29 10:33:21.851198522 -0400
@@ -337,7 +337,7 @@
   >HorizonX, s.
 
   {Copy to video table. This is most timing-critical for a smooth road update}
-  \SYS_RacerUpdateVideoX_40 \sysFn: 40!!
+  {\SYS_RacerUpdateVideoX_40} HorizonAddr $3f- \sysFn: 40!!
   {This SYS call self-repeats for a total of 47 times and is equivalent to:
     [do
       q, X- p. p 4- p=
@@ -430,6 +430,14 @@
 [def #3             #40 #20 #20 #63 #63 #20 #20 #40             #0] Car1=
 [def #2         #40 #20 #20 #20 #40 #40 #20 #20 #20 #40         #0] Car2=
 
+{ FindHorizon }
+[def _FindHorizon=*
+  \SYS_ReadRomDir_v5_80 _sysFn=
+  0 [do 80!! p=
+      $697a _sysArgs0^ [if=0 p ret]  {match 'zi' only!}
+    p loop]
+    ret
+]
 
 {-----------------------------------------------------------------------+
 |}$08a0 \vLR: ret{      RAM page 8                                      |
@@ -465,7 +473,7 @@
   $012e \sysArgs0: {"p"}
   $20   \sysArgs2. {"SegmentY"}
   \invTable q=
-  \SYS_RacerUpdateVideoY_40 \sysFn:
+  {\SYS_RacerUpdateVideoY_40} HorizonAddr $26- \sysFn:
   [do
     q 8?
     Distance+ \sysArgs3. 40!!
@@ -628,7 +636,7 @@
 [def
   push
 
-  \zippedRacerHorizon p= $1000 q=
+  HorizonAddr p= $1000 q=
   [def <p, 250^ [if<>0 1 else 6] p+ p= ret] tmp= {"next p"}
   [do
     p 0? \sysArgs0. tmp!
@@ -649,6 +657,8 @@
 |       Run                                                             |
 +-----------------------------------------------------------------------}
 
+\FindHorizon! HorizonAddr=
+
 Intro!
 
 { Main loop }
